### ISSUE TITLE
'Fix' Async `for` semantics

### ISSUE THREAD

**Issue by bartelink** [fsprojects/FSharp.Control.TaskSeq#129]
At present the current impl of [`for` within `async {` expressions](https://github.com/fsprojects/FSharp.Control.TaskSeq/blob/a488161cfbc132d3c6f6018bb17468513eabb86b/src/FSharp.Control.TaskSeq/AsyncExtensions.fs#L11-L12) raises two concerns for me:
- [`AwaitTaskCorrect`](https://github.com/fsharp/fslang-suggestions/issues/840) semantics would be preferable to promulgating usage of `Async.AwaitTask` in a place that most people will not necessarily even infer that it's in play.
- AIUI `Async.StartAsTask` includes an unnecessary transition to the Thread Pool (with a potential context switch?) which could instead safely be replaced with `Async.StartImmediateAsTask` in this context
- the starting of the child Async does not propagate then continuation token (not surewhether fixing that is possible/required)

**Comment by bartelink** [comment]
Fringe ramblings which you may consider related: I tend to have an [`Async.startImmediateAsTask` helper](https://github.com/jet/equinox/blob/b0fb2e7549afbbee847dd629f31a01c2957740fd/src/Equinox.DynamoStore/DynamoStore.fs#L423-L425) that I try to route the starting of child Tasks through within `async {`s :- 

```
module private Async =

    let startImmediateAsTask ct computation = Async.StartImmediateAsTask(computation, cancellationToken = ct)
```

Aside: the term `Immediate` never really spoke to me; I tended to quickly discount it from consideration. I assume that naming is probably intentional. While I doubt aliasing an established name is likely to pass any scrutiny, perhaps naming such a helper something like `Async.executeTask` might express the intention better? (I'd call it `Async.runAsTask`, but for me the use of `Run` connotes `Task.Run`, i.e. implies an invocation via the Thread Pool)

Related: Potentially would advocate for inclusion of any `startImmediateAsTask` helper in a shims lib alongside #128

**Comment by abelbraaksma** [comment]
Yes, before I wrote this, I wasn't aware of the issues with `Async.AwaitTask`. This can clearly be improved. I'll probably need a test that showcases how this goes off the rails, to ensure the implementation of `AwaitTaskCorrect` (by whatever other name) is, well, correct.

As with #128, we should consider placing these in a separate library. Though for the sake of this one being a no-dependency library by choice, for now we'll just implement it here (and/or copy or link it from whatever other library we conceive).

**Comment by bartelink** [comment]
>This can clearly be improved. I'll probably need a test that showcases how this goes off the rails, to ensure the implementation of AwaitTaskCorrect (by whatever other name) is, well, correct.

Yes, the honoring of Async.CancellationToken aspect to terminate the wait is notably subtle to me and makes covering it well worthwhile.
But by the time you've done that, the canonical impl will likely see some change (and I bet the perf can be improved) 

>Though for the sake of this one being a no-dependency library by choice

I'm in a similar quandary with Equinox - the core lib is tiny and making it depend on another lib only for AwaitTaskCorrect feels like jumping the shark and going leftpadesque. There are 5 other libs that need both TaskSeq and AwaitTaskCorrect. But the bottom line is that having a separate lib, no matter how small, is simply the right thing to do. UNLESS/until it goes into FSharp.Core.... (see conclusion below)

FsKafka is another case where the lib is a 400 line slab that has a stupid other file that dilutes that simplicity and directness thanks to AwaitTaskCorrect

FSharp.AWS.DynamoDB has a copy too, and I really would like for the world to have a way of knowing that a given exe has only one set of AwaitTask semantics in play when reasoning about stack traces and the inevitably hairy cancellation and/or hang issues (see https://github.com/dotnet/fsharp/issues/13165 - a large part of whittling that down was doing stupid busywork diffing to rule that sort of stuff out). (for that lib, the larger issue is that it should probably present a `Task` layer, and then layer its current `Async` API over that - atm Equinox is having to do a redundant layer of Task over Async which then immediately drops to Task in the underlying impl within FSAWSDDB)

Having said all that, I respect the value of not having a package dependency and will likely distribute a copy of the final AwaitTaskCorrect into FSharo.AWS.DynamoDB, FsKafka, Equinox and Propulsion if that's what TaskSeq does (and will argue against making TaskSeq's version public to avoid people havign to self-discipline to not take a depedency on TaskSeq only for AwaitTaskCorrect). But I think I am saying I'm voting for `*.ignore` and `Async.toTask` to go in a single nuget until they can go into FSharp.Core where they should really be for everyone's sake.

**Comment by bartelink** [comment]
> Related: Potentially would advocate for inclusion of any `startImmediateAsTask` helper in a shims lib alongside #128
Addendum to this re https://github.com/fsprojects/FSharp.Control.AsyncSeq/issues/74#issuecomment-1042465992
- An (Async|Task).parallelThrottled would go well with `Async.runAsTask` and/or `(Async|Task|ValueTask).ignore` (forcing specification of a cancellationtoken and a degree of parallelism to be considered when going parallel). At present I have some ugly cases where I am abusing Async.Parallel within code that's Task-based. It might be a stretch for them to go into FSharp.Core for reasons of redundancy as per Don's comments, but having a clear surface area that makes one consider cancellation in all cases, and degreeOfParallelism in cases of going parallel would help in general IMO
- are `iterParallel(Async)?` and `mapParallel(Async)?` missing from the roadmap?
- could consider the notion of a generic `Async.throttle` as a way to constrain parallelism vs having a degreeOfParallism aspect, though a) that's likely less optimal and b) arguably that is less likely to have users end up in good designs
- as it stands I will cart likely cart around a `mapAsyncParallelThrottled` helper as per https://github.com/fsprojects/FSharp.Control.AsyncSeq/issues/74#issuecomment-1034872069 until such time as either that becomes directly available, or there's a `mapAsyncParallel` that I then constrain via having an Async.Throttle and/or Task-awaitable lightweight semaphore wrapper

**Comment by abelbraaksma** [comment]
Thanks for all the input, some work is cut out for me :).

> But I think I am saying I'm voting for *.ignore and Async.toTask to go in a single nuget until they can go into FSharp.Core where they should really be for everyone's sake.

Alternatively: we can take a source dependency instead. That way it can be internal.

**Comment by bartelink** [comment]
Ah, I missed the mention of the fact that parallelism is off the table for now in https://github.com/fsprojects/FSharp.Control.TaskSeq#further-reading-iasyncenumerable - focusing on getting stuff mapped out in the core feature set absolutely makes sense, especially given the non-trivial nature of the ancillary aspects as highlighted by this very issue.

**Comment by abelbraaksma** [comment]
Just FYI: the `StartAsTask` and `StartImmediateAsTask` points that you brought up here escaped my attention somehow. They're now separately under the umbrella of #135. 

This thread can then solely focus on the issues around `AggregateException` and `AwaitTaskCorrect` (needs a better name).

**Comment by bartelink** [comment]
👍 Apologies for mixing the concerns in the first instance. Putting this here as this thread has already run amok; not sure if it's really `TaskSeq` business but it would be good to see it result in either an agreed set of helpers, or a summary document somewhere. Some of this is covered in https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/128, but I'd like to explore the surface area a tiny bit more here first before this reverts to being a simple issue about the `for` impl being corrected...

TL;DR there needs to be a canonical set of helpers somewhere that emphasize:
- not being painful to use with piping
- with good names

`TaskSeq` currently
a) uses/should use the bulk of these
b) exposes some of the others

Below the fold, I have itemised the APIs I believe should be considered for wrapping.

My hope is that we can: 
a) converge on what APIs go into the `TaskEx` signature set so it can go into a fresh issue
b) Close #128 as it is currently confusing about whether it is trying to address `ignore`, `awaitTask`, or the whole lib, pointing to ☝️ 
c) Close this in favour of a fresh issue that links to [`AwaitTaskCorrect`](https://github.com/fsharp/fslang-suggestions/issues/840) and https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/135
 
----

## `ignore`

(Bear with me, making an overall point)
`Async.Ignore` has always been ugly and undiscoverable. While I tend to do `let! _ = <async stuff I want to ignore result of>`, it's commonly the last expression in a function, and having to do `let! _ = <thing I'm wrapping> in ()` is too much. 

=> As appears to the be direction taken in this PR, any given construct should have a correct `ignore` impl, that corrects observes exceptions etc

```
module Async =
    let inline ignore (a : Async<'t>) = Async.Ignore
module Task =
    let inline ignore
module ValueTask =
    let inline ignore
```

## `StartAsTask` / `Start`

- `Start` normally doesn't confuse people - It's clear it's in the thread pool, and cancellation can normally be swept under the rug.
- `StartAsTask` has signature `computation * ?taskCreationOptions *. ?cancellationToken`
    - passing a cancellation token requires `Async.StartAsTask (computation, ?cancellationToken = ct)`
    - the name does not scream thread hop/thread pool
    - passing a cancellation token should not be glossed over, ever; `Async` being replaced with `Task` everywhere does not change the importance of treating the correct propagation of cancellation token as a first class concern that should always be on the table

Common helper:
```
module Async =
    let inline runAsTask ct (a : Async<'t>) = Async.StartAsTask(a, cancellationToken = ct)
```

## `StartImmediateAsTask`

As noted, this should be the default way in which `Async`s are started.
- Because it is not starting a thread, it does not have a `taskCreationOptions` arg so it happens one can use it directly for piping.
- The `immmediate` bit does not align with anything I'm aware of, although that also means it's meaning, once understood, is not ambiguous. I'd be suggesting `execute` as a verb to imply "do it here and now" alongside `run` as a better name igf the precedent had not been established.

My best suggestion is thus to have:

```
module Async =
    let inline startImmediateAsTask ct (a : Async<'t>) = Async.StartImmediateAsTask(a, cancellationToken = ct)
   // MAYBE
    let inline execute ct (a : Async<'t>) = Async.StartImmediateAsTask(a, cancellationToken = ct)
```

## `AwaitTask`

The default impl is baked and wrong, but [adding an overload is both desirable and questionable](https://github.com/fsharp/fslang-suggestions/issues/840#issuecomment-1156311347) 

- this lib has `Async.toTask` and `Task.ofAsync`

```
module Async =
    let inline ofTask (t : Task<'t>) : Async<'t> = AwaitTaskCorrect t
module Task =
    let inline toAsync (t : Task<'t>) : Async<'t> = Async.ofTask t
```

## Async.Parallel

The degree of parallelism parameter was added late in the game, but is critical - adding an arbitrary number of items to the threadpool should be a very questionable act.
- pipelining is painful
- having Throttled in the name is pretty well established https://www.compositional-it.com/news-blog/improved-asynchronous-support-in-f-4-7
- before v FSharp.Core v 6.0.6, [there was a stack overflow bug that can tear down the process](// https://github.com/dotnet/fsharp/issues/13165) if >1200 items are started with a throttle and cancellation is triggered quickly

```
module Async =
    let parallelThrottled dop computations =
        Async.Parallel(computations, maxDegreeOfParallelism = dop)
```

Further puzzlers in this space (which probably mean this should be excluded from the surface area of any set of helpers):
- A common case is to use this to run but await failure of multiple `Async<unit>` tasks, having to do an `Async.Ignore<unit[]>` is ugly for that
- How do you swap back/forth from that to `Task`, considering cancellation tokens and unwrapping AggregateException

**Comment by abelbraaksma** [comment]
Maybe you can copy that whole text as a new issue? It feels like it actually belongs in #128, no? Basically the whole idea of agreeing on a surface area for Task/Async, likely in its own repo and package.

Maybe I should just setting one up and we can continue there?

**Comment by bartelink** [comment]
@dsyme For avoidance of doubt, the PRs being stacked here are questionable - generic small task helpers in this lib that are orthogonal in nature

I'm considering getting around to doing the AwaitTaskCorrect in FSharp.Core which is the real answer here. IIRC there are also other AwaitTasks beside the one it found around here, and those need to be consistent for sanity too.
